cpp/ql/test/experimental/query-tests/Security/CWE/CWE-703/semmle/tests/test.cpp

```diff
@@ -0,0 +1,26 @@
+typedef unsigned int size_t;
+void clean();
+
+
+void funcTest1()
+{
+   throw ("my exception!",546);  // BAD
+}
+
+void DllMain()
+{
+  try { throw "my exception!"; } // BAD
+  catch (...) {  }
```

Comment:
Why is this case bad?  The exception appears to be caught by the surrounding `try...catch`.

Reply:
the exception handler is empty, which means that the body of the library will not be unloaded from the process's address space.

Reply:
I could see that if the exception was re-thrown in the `catch` block, but in this case it appears to be handled (admittedly by doing nothing - perhaps the developer knows this exception can be safely ignored?) and normal execution resumes.

Reply:
As far as I understand.
The loader expects a return from the `dllmain` function.
and if it doesn't, it doesn't unload the library.

technically, any processing that does not include "return" can be considered bad. but I decided to limit myself to a narrower detection.

Reply:
> technically, any processing that does not include "return" can be considered bad. but I decided to limit myself to a narrower detection.

But a `void` function does not need an explicit `return` statement as far as I know.
So after the exception is handled in the catch block (by doing nothing) this should just return without any problems and the library should be unloaded successfully.

Reply:
Throwing an exception that escapes DllMain is bad. Not only does the library not get unloaded, but if something inside the Windows loader implementation does attempt to handle the exception, it won't treat it as a C++ exception, and therefore the exception object itself won't be destructed properly. This is all just a general consequence of throwing an exception from a function whose contract doesn't allow it to throw exceptions.

Throwing an exception that is also caught and handled within DllMain should be fine. However, a DllMain implementation must return a BOOL (typedef for int). The test case above an example of bad code, but not because of any usage of exceptions. It's just bad because it doesn't return a BOOL from DllMain. If the return type were changed to BOOL, the compiler would warn if it did not return a value.

Reply:
@ihsinme do you intend to make any changes addressing this or leave the query as it is?

Reply:
@geoffw0
I would leave it as is to better match the name of the request.
but if you think otherwise, I am willing to change to detect all dllmain from which there is an exit, without returning a boolean value.